File: sdks/node/src/pagination.ts
/**
 * HookSniff SDK — Auto-pagination iterator
 * Provides `for await` pagination over list endpoints.

 * Usage:
 *   for await (const ep of hooksniff.endpoints.list()) {
 *     console.log(ep.id);
 *   }
 */

export interface ListResponse<T> {
  data: T[];
  iterator?: string | null;
  done?: boolean;
}

export interface PaginationOptions {
  limit?: number;
  iterator?: string | null;
}

/**
 * Creates an async iterator that automatically fetches the next page
 * until all items are consumed.
 */
export function paginatedIterator<T, O extends PaginationOptions = PaginationOptions>(
  fetchPage: (options?: O) => Promise<ListResponse<T>>,
  options?: O
): AsyncIterable<T> {
  return {
    [Symbol.asyncIterator]() {
      let currentIterator: string | null | undefined = options?.iterator ?? null;
      let done = false;
      let buffer: T[] = [];
      let bufferIndex = 0;

      return {
        async next(): Promise<IteratorResult<T>> {
          // If buffer has items, return next
          if (bufferIndex < buffer.length) {
            return { value: buffer[bufferIndex++], done: false };
          }

          // If we're done, return
          if (done) {
            return { value: undefined as unknown as T, done: true };
          }

          // Fetch next page
          const pageOptions = {
            ...options,
            iterator: currentIterator,
          } as O;

          const response = await fetchPage(pageOptions);

          buffer = response.data ?? [];
          bufferIndex = 0;

          // Check if there are more pages
          if (response.done === true || !response.iterator || buffer.length === 0) {
            done = true;
          } else {
            currentIterator = response.iterator;
          }

          // If this page was empty, we're done
          if (buffer.length === 0) {
            return { value: undefined as unknown as T, done: true };
          }

          return { value: buffer[bufferIndex++], done: false };
        },
      };
    },
  };
}

File: sdks/python/hooksniff/pagination.py
"""
HookSniff SDK — Auto-pagination

Provides automatic pagination over list endpoints.

Usage:
    for ep in hs.endpoints.list_all():
        print(ep["url"])
"""

from __future__ import annotations

from typing import Any, Callable, Generator, Generic, Optional, TypeVar

T = TypeVar("T")


class Page(Generic[T]):
    """Represents a single page of results."""

    def __init__(self, data: list[T], iterator: str | None = None, done: bool = False):
        self.data = data
        self.iterator = iterator
        self.done = done


def paginate(
    fetch_page: Callable[..., dict[str, Any]],
    limit: int | None = None,
    iterator: str | None = None,
) -> Generator[dict[str, Any], None, None]:
    """
    Auto-paginate through all items from a list endpoint.

    Args:
        fetch_page: Function that fetches a single page
        limit: Items per page
        iterator: Starting iterator

    Yields:
        Individual items from all pages
    """
    current_iterator = iterator
    done = False

    while not done:
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if current_iterator is not None:
            params["iterator"] = current_iterator

        response = fetch_page(**params)
        data = response.get("data", [])

        for item in data:
            yield item

        if response.get("done") is True or not response.get("iterator") or not data:
            done = True
        else:
            current_iterator = response["iterator"]


def collect_all(
    fetch_page: Callable[..., dict[str, Any]],
    limit: int | None = None,
) -> list[dict[str, Any]]:
    """
    Collect all items from a paginated endpoint into a list.

    Args:
        fetch_page: Function that fetches a single page
        limit: Items per page

    Returns:
        List of all items
    """
    return list(paginate(fetch_page, limit=limit))
